File: 0430-flatten-a-multilevel-doubly-linked-list/0430-flatten-a-multilevel-doubly-linked-list.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head) {
        stack<Node*>stk;
        if(!head)return nullptr;
        stk.push(head);
        Node*prev=nullptr;
        while(!stk.empty()){
            Node*curr=stk.top();
            stk.pop();
            if(prev){
                prev->next=curr;
                curr->prev=prev;
            }
            if(curr->next){
                stk.push(curr->next);
            }
            if(curr->child){
                stk.push(curr->child);
                curr->child=nullptr;
            }
            prev=curr;
        }
        return head;
    }
};

File: 0430-flatten-a-multilevel-doubly-linked-list/0430-flatten-a-multilevel-doubly-linked-list.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};
*/

class Solution {
    public Node flatten(Node head) {
        Stack<Node>st=new Stack<>();
        if(head==null)return null;
        st.push(head);
        Node prev=null;
        while(!st.isEmpty()){
            Node curr=st.peek();
            st.pop();
            if(prev!=null){
                prev.next=curr;
                curr.prev=prev;
            }
            if(curr.next!=null){
                st.push(curr.next);
            }
            if(curr.child!=null){
                st.push(curr.child);
                curr.child=null;
            }
            prev=curr;
        }
        return head;
    }
}